packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+        // bail on a rest element
+        if (property.type === AST_NODE_TYPES.RestElement) {
+          return;
+        }
+
+        const valueType = services.getTypeAtLocation(property.value);
+
+        // bail if misses; this is a type-error
+        if (tsutils.isIntrinsicErrorType(valueType)) {
+          return;
+        }
+
+        const memberKey = getStaticMemberAccessValue(property, context);
+
+        // collect dynamic keys which we failed to statically analyzed
+        if (memberKey === undefined) {
+          dynamicProperties.add({ property });
+          continue;
+        }
+
+        remainingProperties.set(memberKey, { property });
+      }
+
+      // remove used type members, report on anything that's unused
+      for (const member of typeAnnotation.members) {
+        if (member.type === AST_NODE_TYPES.TSIndexSignature) {
+          checkIndexSignature(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+
+        if (
+          member.type === AST_NODE_TYPES.TSMethodSignature ||
+          member.type === AST_NODE_TYPES.TSPropertySignature
+        ) {
+          checkStaticMember(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+      }
+    }
+
+    function checkTuple(
+      param: TSESTree.ArrayPattern,
+      typeAnnotation: TSESTree.TSTupleType,
+    ): void {
+      for (const [index, member] of typeAnnotation.elementTypes.entries()) {
+        // bail on rest type `[...string[]]`
+        if (member.type === AST_NODE_TYPES.TSRestType) {
+          return;
+        }
+
+        const property = param.elements.at(index);
+
+        if (property === undefined) {
+          reportOnMember(member, { type: 'key', key: String(index) });
+          continue;
+        }
+
+        if (
+          // bail in case of `[, ...]`
+          property == null ||
+          // bail on a rest element
+          property.type === AST_NODE_TYPES.RestElement
+        ) {
+          return;
+        }
+
+        checkParam(property, member);
+      }
+    }
+
+    /**
+     * Checks if an index signature property is covered by at least one destructure property.
+     */
+    function checkIndexSignature(
+      member: TSESTree.TSIndexSignature,
+      remainingProperties: RemainingProperties,
+      dynamicProperties: DynamicProperties,
+    ) {
+      const indexParameterType = getConstrainedTypeAtLocation(
+        services,
+        // an index signature must have exactly one parameter, having more
+        // is valid syntax, but invalid typescript
+        member.parameters[0],
+      );
+
+      if (
+        indexSignatureHasMatchingKey(
+          indexParameterType,
+          dynamicProperties,
+          remainingProperties,
+        )
+      ) {
+        return;
+      }
+
+      // Template literal index signatures count as used if any remaining property's type is the
+      // same, this may miss some unused template literal index signatures
```

Comment:
> this may miss some unused template literal index signatures

[Docs] What do you mean by this? (asking not just rhetorically, this should be in the docs if it's relevant)